KAFKA-3912: Query local state stores

dguy opened the pull request:
@guozhangwang @enothereska please review

dguy committed:
provide access to local state-stores

dguy committed:
add all and range to ReadOnlyKeyValueStore

dguy committed:
Merge pull request #1 from dguy/expose-local-stores

Expose Local State Stores

dguy committed:
throw InvalidStateStoreException when Store not open or no stores available

dguy committed:
Introduce and use QueryableStoreType for locating/querying stores of different types

dguy committed:
revert accidental changes due to renames

dguy committed:
add some javadoc

dguy committed:
fix RocksDB open check. JavaDoc

dguy committed:
extract common interface for store providers

dguy committed:
Merge pull request #3 from dguy/alternative-local-api-spike

Query Local State Stores

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreTypes.java:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Provides access to the {@link QueryableStoreType}s provided with KafkaStreams. These
+ * can be used with {@link org.apache.kafka.streams.KafkaStreams#getStore(String, QueryableStoreType)}
+ * To access and query the {@link StateStore}s that are port of a Topology
```
typo: port->part

enothereska commented on streams/src/main/java/org/apache/kafka/streams/QueryableStoreProvider.java:
```diff
@@ -0,0 +1,56 @@
+        if (allStores.isEmpty()) {
+            return null;
+        }
+        return queryableStoreType.create(
```
This part ignores the just computed `allStores` right? Was the `allStores` code just to check if it's null or did you intend to use it further?

dguy replied:
It is just to check that a Store with the given name and QueryableType exists before creating the Composite type

enothereska commented:
The new interfaces LGTM. Great tests!

dguy committed:
fix type in javadoc. fix compilation error

dguy committed:
check is streams app is running in getStore call

dguy commented:
@mjsax @miguno

dguy commented:
Just realised that RocksDBWindowStore is going to need some locking/synchronization around the creating and cleaning up of segments. Looking now to try and find the least intrusive way

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/QueryableStoreProvider.java:
```diff
@@ -0,0 +1,56 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.kafka.streams;
```
Why we need to keep this class in top-level streams package?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreType.java:
```diff
@@ -0,0 +1,47 @@
+
+    /**
+     * Create an instance of T (usually a facade) that developers can use
+     * to query the Underlying {@StateStore}s
```
Underlying -> underlying

{@StateStore} -> {@link StateStore}, and I saw some other places with similar tags.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreTypes.java:
```diff
@@ -0,0 +1,89 @@
+        return new WindowStoreType<>();
+    }
+
+    public static abstract class QueryableStoreTypeMatcher<T> implements QueryableStoreType<T> {
```
This can be private.

EDIT: actually nvm, I saw that customized stores from user can use this as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreTypes.java:
```diff
@@ -0,0 +1,89 @@
+    private static class KeyValueStoreType<K, V> extends
+                                                 QueryableStoreTypeMatcher<ReadOnlyKeyValueStore<K, V>> {
+        KeyValueStoreType() {
+            super(ReadOnlyKeyValueStore.class);
```
Maybe I mis-understand something, but should this be `KeyValueStore.class`?

EDIT: okay, nvm again.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreTypes.java:
```diff
@@ -0,0 +1,89 @@
+    private static class WindowStoreType<K, V> extends QueryableStoreTypeMatcher<ReadOnlyWindowStore<K,
+            V>> {
+        WindowStoreType() {
+            super(ReadOnlyWindowStore.class);
```
Ditto here.

dguy replied:
We want `ReadOnlyWindowStore.class` as it restricts the API to only `fetch`. `WindowStore extends ReadOnlyWindowStore`

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/UnderlyingStoreProvider.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Provides access to {@link org.apache.kafka.streams.processor.StateStore}s accepted
+ * by {@link QueryableStoreType#accepts(StateStore)}
+ * @param <T>
```
This should be moved to `getStores` java doc string, as the class itself is not templated.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyStore.java:
```diff
@@ -0,0 +1,136 @@
+ * @param <K> key type
+ * @param <V> value type
+ */
+public class CompositeReadOnlyStore<K, V> implements ReadOnlyKeyValueStore<K, V> {
```
Rename to CompositeReadOnlyKeyValueStore?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreTypes.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private static class WindowStoreType<K, V> extends QueryableStoreTypeMatcher<ReadOnlyWindowStore<K,
+            V>> {
```
nit: put in the same line?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/UnderlyingStoreProvider.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.List;
+
+/**
+ * Provides access to {@link org.apache.kafka.streams.processor.StateStore}s accepted
```
The java doc string can be improved here: we can state that this is to wrap multiple underlying {@link StateStoreProvider}s as a single provider.

Also the name is a bit awkward to me, could we rename it to sth. like `WrappingStoreProvider`?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TestKeyValueStore.java:
```diff
@@ -0,0 +1,146 @@
+import java.util.NoSuchElementException;
+import java.util.TreeMap;
+
+class TestKeyValueStore<K, V> implements KeyValueStore<K, V> {
```
Could we just rename to "InMemoryKeyValueStore"?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStore.java:
```diff
@@ -0,0 +1,75 @@
+    public WindowStoreIterator<V> fetch(final K key, final long timeFrom, final long timeTo) {
+        final List<ReadOnlyWindowStore<K, V>> stores = provider.getStores(storeName, windowStoreType);
+        for (ReadOnlyWindowStore<K, V> windowStore : stores) {
+            final WindowStoreIterator<V> result = windowStore.fetch(key, timeFrom, timeTo);
```
This iterator actually wraps multiple underlying segment which could be deleted / replaced concurrently. We need to consider how to handle this.

dguy replied:
@guozhangwang in this case perhaps we should change the `RocksDBWindowStoreIterator` to take an array of `Segment` rather than an array of `KeyValueIterator` it would also need the `from` and `to`. Given that the, `RocksDBWindowStoreIterator` can check if the `Segment` `isOpen` before issuing the range query against the segment.

So the constructor would look something like:
`RocksDBWindowStoreIterator(StateSerdes<?, V> serdes, Bytes from, Bytes to, Segment[] segments)`

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyStore.java:
```diff
@@ -0,0 +1,136 @@
+        final NextIteratorFunction<K, V> nextIteratorFunction = new NextIteratorFunction<K, V>() {
+            @Override
+            public KeyValueIterator<K, V> apply(final ReadOnlyKeyValueStore<K, V> store) {
+                return store.all();
```
Should we check here if the store is already closed?

Also, does RocksDB handles the case where an iterator is still open while its underlying store is closed? If it does not handle this case then we should do it.

dguy replied:
Interestingly, i ran a test yesterday and the iterator still works even after the store has been closed. I'm not sure for how long. But yes, we should check here if the store is closed.

dguy replied:
Actually the isOpen check happens in `StreamThreadStateStoreProvider.getStores(..)` however there probably should be an additional `isOpen` check in the iterator

dguy replied:
And looking into it further, there is an additional check in the `RocksDBStore` when `all` or `range` is called. if the store is not `open` an exception will be thrown - which i think is valid in this case. It would mean a re-balance is in process.

dguy committed:
address comments from @guozhang. Thread-safety in MemoryLRU, RocksDBStore & RocksDBWindowStore

dguy committed:
Change RocksDBWindowStoreIterator to check if segments are open

dguy commented:
@guozhangwang - addressed all of your comments and made some changes for concurrent access. Things I really need another pair of eyes on
`RocksDBWindowStore`, `RocksDBStore`, `MemoryLRUCache`
Thanks

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/StateStore.java:
```diff
@@ -58,4 +58,10 @@
      * @return  {@code true} if the storage is persistent&mdash;{@code false} otherwise
      */
     boolean persistent();
```
To be consistent, we should perhaps rename `persistent()` to `isPersistent()`?

dguy replied:
I should probably rename `isOpen` to `open`. The style seems to be to drop the `get` and `is` prefixes

ijuma replied:
I think it's important to maintain clarity independently of conventions. `getName()` and `name()` are both clear and the former has additional verbosity. `open` is quite ambiguous IMO. If I saw a method name like that, I would expect it to `open` something instead of returning whether it is open. I am not sure if the `is` prefix is discouraged in the same way as the `get` prefix for this reason.

dguy replied:
ok - i agree with that @ijuma. Thanks for clarification

nehanarkhede replied:
Agree w/ @ijuma here

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/StateStore.java:
```diff
@@ -58,4 +58,10 @@
+     * Is this store open
+     * @return {@code true} if the store is open
+     */
+    boolean isOpen();
```
What's the meaning of being open (`isOpen() == true`)?  Does it mean the store is ready for writing/reading?

miguno commented on streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreTypes.java:
```diff
@@ -0,0 +1,88 @@
+        public QueryableStoreTypeMatcher(Class matchTo) {
+            this.matchTo = matchTo;
+        }
+        @SuppressWarnings("unchecked")
```
Nit: Newline before.

miguno commented on streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java:
```diff
@@ -0,0 +1,52 @@
+    KeyValueIterator<K, V> range(K from, K to);
+
+    /**
+     * Return an iterator over all keys in the database. This iterator MUST be closed after use.
```
"in the database" -> "in this store" or "in this state store"

miguno commented on streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @return An iterator of all key/value pairs in the store.
+     */
+    KeyValueIterator<K, V> all();
```
Question: Should we follow the naming of Java collections?

miguno commented on streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java:
```diff
@@ -0,0 +1,52 @@
+     * @return The iterator for this range.
+     * @throws NullPointerException If null is used for from or to.
+     */
+    KeyValueIterator<K, V> range(K from, K to);
```
For `range` and `all`: Should we require a specific contract?  For example, do we want to define the behavior of the iterator if the underlying state store is being modified (e.g. what happens if, after the iterator is returned, a key is removed from the table)?  Do we assume a snapshot view?

miguno replied:
What's the ordering for `K` when we query for a range?  In Scala you'd have e.g. an implicit `Ordering[K]`.  Should we use a  `Comparable<K>` somewhere here?

mjsax replied:
I had the same thought about the ordering...

About `range/all` -- I think a snapshot view might be quite expensive (need to copy the whole local KV-store and block all updates as long as copying is done -- not sure if we want to support this?). On `next` we could just step to the next available key in the range. Handling updates to the key range might be tricky -- also if current key got deleted before `next` is called.

dguy replied:
I don't think we can guarantee Ordering in this case as the range of keys will possibly be over multiple Stores, i.e, when you are querying you are actually using a Facade wrapping the underlying state stores - i don't think we want want to bring all the keys in to memory and sort them.

Taking snapshots could be very expensive - not sure we want to do this for every query. RocksDB does have built in support for this, though.

Handling updates/deletions etc has to be dealt with at the `StateStore` level, i.e, each store type will have different semantics. However, we probably should state that the returned iterator needs to be safe from Concurrent Modifications. Two examples.

**RocksDBStore**

```
store.put(1, "one");
store.put(2, "two");
store.put(3, "three");
store.put(4, "four");
final KeyValueIterator<Integer, String> range = store.range(1, 5);
store.delete(4);
store.put(2, "eight");
store.flush();
while (range.hasNext()) {
        System.out.println(range.next());
 }
```

Results:

```
KeyValue(1, one)
KeyValue(2, two)
KeyValue(3, three)
KeyValue(4, four)
```

**ConcurrentHashMap**

```
final ConcurrentHashMap<Integer, String> map = new ConcurrentHashMap<>();
map.put(1, "one");
map.put(2, "two");
map.put(3, "three");
map.put(4, "four");

final Iterator<Map.Entry<Integer, String>> iterator = map.entrySet().iterator();
map.remove(4);
map.put(2, "eight");
while (iterator.hasNext()) {
   System.out.println(iterator.next());
}
```

Results:

```
2=eight
1=one
3=three
```

I know - `ConcurrentHashMap` doesn't have a `range(..)` method, but it demonstrates that the iterator is safe. HashMap would throw a `ConcurrentModificationException`

miguno replied:
Thanks @dguy.  That's exactly what I was aiming at -- i.e. we should clarify the contract of `range` and `all`, even if that contract would be "there's no guaranteed behavior".  The user should at least know what she's in for.

miguno commented on streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyWindowStore.java:
```diff
@@ -0,0 +1,35 @@
+     *
+     * @return an iterator over key-value pairs {@code <timestamp, value>}
+     */
+    WindowStoreIterator<V> fetch(K key, long timeFrom, long timeTo);
```
We should clarify exclusive/inclusive for timeFrom and timeTo.

dguy replied:
I think we need to clarify it in our code first. At the moment, the `RocksDBRangeIterator` is inclusive of both. However, the `MemoryNavigableLRUCache.CacheIterator` is inclusive of `timeFrom` and exclusive `timeTo` - we should agree what it should be and fix @guozhangwang

guozhangwang replied:
Yeah I think it should be inclusive at both ends, as we stated in the interface:

```
/**
     * Get an iterator over a given range of keys. This iterator MUST be closed after use.
     *
     * @param from The first key that could be in the range
     * @param to The last key that could be in the range
     * @return The iterator for this range.
     * @throws NullPointerException If null is used for from or to.
     */
    KeyValueIterator<K, V> range(K from, K to);
```

And `MemoryNavigableLRUCache.range()` has a bug that needs to be fixed.

dguy replied:
Thanks for clarifying. I'll change the `MemoryNavigableLRUCache.range()` to be inclusive at both ends

miguno commented on streams/src/main/java/org/apache/kafka/streams/state/StateStoreProvider.java:
```diff
@@ -0,0 +1,39 @@
+     * @param storeName             name of the store
+     * @param queryableStoreType    filter stores based on this queryableStoreType
+     * @param <T>                   The type of the Store
+     * @return  List of the instances of the store in this topology. Empty List if not found
```
IIRC sometimes we return `null` in such cases, here we return an empty list.  Just asking.  I prefer to not return `null` to minimize the likelihood of downstream NPEs, but perhaps we favor consistency.

dguy replied:
IMO the right thing to do is return an empty list. Unless someone strongly disagrees with this i'm going to leave it that way

miguno replied:
Sounds good to me!

miguno commented on streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java:
```diff
@@ -0,0 +1,52 @@
+ * @param <V> the value type
+ */
+@InterfaceStability.Unstable
+public interface ReadOnlyKeyValueStore<K, V> {
```
What about the thread safety (requirements, assumptions, ...) of these new interfaces?

miguno commented on streams/src/test/java/org/apache/kafka/streams/state/NoOpWindowStore.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    public String name() {
+        return null;
```
I know it's a test class, but IIRC a state store name must not be null? (I might be wrong.)

miguno commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java:
```diff
@@ -0,0 +1,169 @@
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class CompositeReadOnlyKeyValueStoreTest {
```
Nice tests!

miguno commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -0,0 +1,297 @@
+        final Set<KeyValue<String, Long>> countState = new TreeSet<>(stringLongComparator);
+
+        final long timeout = System.currentTimeMillis() + 30000;
+        while (joinThisState.size() < 5 &&
```
At least we try to not abuse Thread.sleep() ;-)

miguno commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStoreTest.java:
```diff
@@ -0,0 +1,115 @@
+    public void before() {
+        stubProviderOne = new StateStoreProviderStub();
+        stubProviderTwo = new StateStoreProviderStub();
+        undelyingWindowStore = new ReadOnlyWindowStoreStub<>();
```
Typo: "undelyingWindowStore" -> "underlyingWindowStore"

miguno commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStoreTest.java:
```diff
@@ -0,0 +1,115 @@
+        undelyingWindowStore = new ReadOnlyWindowStoreStub<>();
+        stubProviderOne.addStore(storeName, undelyingWindowStore);
+
+        otherUnderylingStore = new ReadOnlyWindowStoreStub<>();
```
Typo: "otherUnderylingStore" -> "otherUnderlyingStore"

miguno commented on streams/src/test/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java:
```diff
@@ -0,0 +1,146 @@
+import java.util.NoSuchElementException;
+import java.util.TreeMap;
+
+class InMemoryKeyValueStore<K, V> implements KeyValueStore<K, V> {
```
Question: Would it help users if `InMemoryKeyValueStore` was under `src/main/` rather than `src/test/`?

dguy replied:
It is just for testing so i wouldn't want to put it under `src/main`

guozhangwang replied:
Could we directly use the `InMemoryKeyValueStore.MemoryStore` here?

miguno commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreTest.java:
```diff
@@ -70,6 +72,45 @@
+        windowStore.put(1, "four");
+
+        // should only have 2 values as the first segment is no longer open
+        assertEquals(new KeyValue<>(60000L, "two"), iterator.next());
```
Where are `60000L` and `120000L` coming from?

dguy replied:
first window starts at 0.
second window starts at 60000.
third window starts at 120000.
i.e., first currentTime is 0. segmentSize is 60000

miguno replied:
Ah, so it's from `RocksDBWindowStoreTest#MIN_SEGMENT_INTERVAL`?

dguy replied:
Yep - `segmentSize` in the test is initialized as `RocksDBWindowStoreTest#MIN_SEGMENT_INTERVAL`

miguno commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -0,0 +1,297 @@
+        streamsConfiguration.put(StreamsConfig.ZOOKEEPER_CONNECT_CONFIG, CLUSTER.zKConnectString());
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG,
+                                 TestUtils.tempDirectory(new File("/tmp").toPath(), "qs-test")
```
Question:  Will this temp directory be cleaned/purged in between test runs?

miguno replied:
And same question for the other uses of `TestUtils.tempDirectory` in this PR.

dguy replied:
Yes - it is deleteOnExit and there is a shutdown hook.

ijuma replied:
Why not simply: `TestUtils.tempDirectory("qs-test")`? That will use the OS's default temp directory (more concise and better behaviour IMO).

dguy replied:
It was to do with the way that `IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);`, i.e., it only clears stuff explicitly under /tmp (not sure why, but that is what it does), but that said it isn't really required anymore as it is using a random dir

ijuma replied:
Thanks for the explanation. It seems like `purgeLocalStreamsState` should really be using `java.io.tmpdir` instead of `/tmp` if it wants to have that safety net.

dguy replied:
Yes - agree. I'll create a JIRA for it as it will require changing a bunch of other tests that are not part of this PR.

ijuma replied:
Sounds good.

miguno commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -0,0 +1,297 @@
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
```
Ah, I see that the state directory is being purged here.  But I think this @After-approach may not work if the test crashes or the JVM terminates before `shutdown` is being called, right? (e.g. due to Jenkins woes) I think it would be safer to purge the state dir prior to the run.

dguy replied:
That doesn't really work. The directory is a tmp directory and is random, the next call in @Before will have a different directory. The worst that happens is that a random directory stays under /tmp

miguno replied:
Good point, thanks for clarifying.

miguno commented on streams/src/test/java/org/apache/kafka/streams/state/internals/QueryableStoreProviderTest.java:
```diff
@@ -0,0 +1,129 @@
+public class QueryableStoreProviderTest {
+
+    private final String keyValueStore = "key-value";
+    private final String windowStore = "window-store";
```
What if both stores have the same name? (but different types)

dguy replied:
Then one of them would fail before we get to this point. There is already a check in the `TopologyBuilder` to make sure you can't add duplicate (named) state stores.

miguno commented:
Thanks for the big chunk of work, @dguy!

dguy committed:
remove get prefix based on KIP feedback from Jay

dguy committed:
remove unused iterator in RocksDBWIndowStore

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -217,4 +228,24 @@ public void setUncaughtExceptionHandler(Thread.UncaughtExceptionHandler eh) {
+        return queryableStoreProvider.getStore(storeName, queryableStoreType);
+    }
+
+    private void validateIsRunning() {
```
Why is this check required? After given the state handle to the user, she might close the application and use the store... (ie, same problem).
Why extra method for check? Only used once.

dguy replied:
Why would you return a Store if the stream wasn't started? If the application is closed, the subsequent use of the store will fail as it would be closed and will throw an `InvalidStateStoreException`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java:
```diff
@@ -81,23 +72,6 @@
-     */
-    KeyValueIterator<K, V> all();
-
-    /**
```
Why no move `approximateNumEntries` to `ReadOnlyKeyValueStore` ?

dguy replied:
Thanks for pointing out - i've added it.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/StateStoreProvider.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.List;
+
+/**
+ * A StateStoreProvider provides access to {@link StateStore}s that have been created
```
Nit: `{@link StateStoreProvider}`

dguy committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-3912

dguy committed:
merge trunk and fix compilation error

dguy committed:
address some PR comments

dguy committed:
fix typo in javadoc

dguy committed:
rename StateStore.isOpen to StateStore.open

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -336,22 +371,34 @@ public V delete(K key) {
+                flushCache();
 
-        return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
+            return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
```
After the iterator is returned, the lock is released, then the stream thread can still close the RocksDB store while iterator is not closed yet, right?

An alternative approach I was thinking, is to keep track of a volatile "RocksObjectCount" for all associated objects, including RocksIterators / WriteBatch / etc, and let the close() call to only proceed when the count becomes 0. Note for WriteBatch since it is only going to be created by the same stream thread it does not actually matter, but just in case it will be created by different threads in the future.

But there is still a race condition if the alternative approach is not carefully implemented such that after the query thread has called `validateStoreOpen()` in `range()`, but before the iterator was created, streams thread can kicks in and delete the store. So we need to increment the count first before checking if store is closed.

dguy replied:
I did start down that path, but I opted against it as I was concerned about callers that grab an iterator and take a long time to iterate or never close it - in this case the `RocksDBStore` wouldn't be able to be closed until the iterator completes, which means rebalances have the potential to become blocked. 
So, i figured that if the store gets closed a call to next will throw an exception. IMO this is perfectly valid. Thoughts? @guozhangwang

dguy replied:
Interestingly, @guozhangwang, the iterator still works even after the store has been closed. I wrote the following:

```
public void shouldSeeWhatHappensWhenAstoreIsClosedWithAnOpenIterator() throws Exception {
        KeyValueStoreTestDriver<Integer, String> driver = KeyValueStoreTestDriver.create(Integer.class, String.class);
        KeyValueStore<Integer, String> store = createKeyValueStore(driver.context(), Integer.class, String.class, false);
        final int max = 100000;
        for (int i = 0; i < max; i++) {
            store.put(i, Integer.toString(i));
        }
        final KeyValueIterator<Integer, String> range = store.range(0, max);
        store.close();
        Thread.sleep(60000);
        while (range.hasNext()) {
            System.out.println(range.next());
        }
    }
```

I've ran this multiple times and even with the 1 minute sleep - it still iterates over the 100k entries without any exceptions.

guozhangwang replied:
Fair enough. Thanks for the validation.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java:
```diff
@@ -105,14 +100,28 @@ public boolean persistent() {
     public void put(K key, V value) {
-        this.map.put(key, value);
-        this.keys.add(key);
+        lock.writeLock().lock();
```
Is it equally safe to make the `map` as a concurrent linkedhashmap, and remove the read lock on `get`? Also if we do this do we still need to lock on `MemoryNavigableLRUCache.toTreeMap`?

dguy replied:
A concurrent linked hashmap doesn't exist.

enothereska replied:
@dguy you'll need a lock for `putIfAbsent` too since the individual locks in put/get are not sufficient (e.g., value could change in between get and put).

ijuma replied:
There is https://github.com/ben-manes/concurrentlinkedhashmap. Guava and Caffeine (Java 8 required) also have Cache implementations with the same underlying behaviour.

dguy replied:
Good point @enothereska  - thanks.
@ijuma - i'm aware of the Guava impls, but i don't think we really want to add Guava as a dependency (unless it already is)?  The ben-manes/concurren... might be worth checking out if people aren't opposed to it?

ijuma replied:
Yeah, I don't know if we do (depends on whether locking is a bottleneck here). And even if we did, it makes sense to do that separately instead of this PR.

ben-manes replied:
CLHM was the baseline algorithm for Guava, though is much faster due to leaving G before optimizing the port. If you later investigate this in-depth, I'd suggest Caffeine now since it includes a superior eviction policy and tons of features. Cheers.

ben-manes replied:
I just read through `MemoryLRUCache`. It is not thread-safe and will corrupt itself because a read causes a mutation of the LRU history. (I made the same mistake early in my career when fixing performance problems leading to exploring caching in-depth, so its an easy oversight to make)

A read/write lock is a very expensive mechanism and most often the incorrect lock type to use. For short critical sections it is more expensive than an exclusive lock. By using a `ReentrantLock` or `synchronized` you'll have both correctness and higher performance.

As is, I strongly urge you to correct this before merging. You don't have to use a caching library, but the code is very broken.

ijuma replied:
Thanks for the review and information @ben-manes. We don't yet require Java 8 and hence why Caffeine is probably not an option for now, but definitely something to consider for the future. I agree that simply using synchronized for now is better than the read write lock.

dguy replied:
Thanks @ben-manes - i had overlooked the fact that gets would be mutating the internal linked list.

dguy committed:
Address PR comments. rename StateStore.open back to StateStore.isOpen. move approximateNumEntries to ReadOnlyKeyValueStore

guozhangwang commented:
The concurrent control in RocksDBStore / RocksWindowStore / MemoryLRUCache look good to me.

Two more things that we can probably do in parallel: while working on a small example leveraging this new API for querying the state stores, also consider writing the documentation sections for this new API (we will incorporate them into the new feature of queryable state store story later); and it'll be great if we can also do some benchmarks with / without the querying with your example code to illustrate the amount of overheads in terms of reduced throughput.

That being said, I think we do not need to make this PR perfect before merging it, though it still needs to wait for the KIP to be adopted. Maybe you can start the voting thread now as there are not too many questions in the discussion thread? After that and all current comments being addressed we can probably merge as is, and come back to revisit or even further change the APIs when we realize issues while writing the example code / docs.

dguy committed:
activeTasks in StreamThread needs to be a ConcurrentHashMap for QS

enothereska commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -237,7 +237,14 @@ public void setUncaughtExceptionHandler(Thread.UncaughtExceptionHandler eh) {
     }
 
+    private void validateIsRunning() {
+        if (state != RUNNING) {
```
@dguy the RUNNING flag is set when the threads are started, but it could be that tasks and their state stores are not initialised yet. Will that create problems?

dguy replied:
The only problem it will cause is that you might not find the store if you called `stores(...)` immediately after `start`.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -490,16 +524,22 @@ public void close() {
             return;
         }
         open = false;
```
Shouldn't the lock be taken before any of these values are changed?

dguy replied:
good point - thanks

enothereska commented:
@dguy I'm wondering if it's possible to add a test or two on concurrency (querying while someone is inserting)? I only ask since you know how to write excellent tests.

dguy committed:
get a WriteLock in MemoryLRUCache.putIfAbsent

dguy committed:
lock entire RocksDBStore.close

dguy committed:
fix merge conflicts

dguy committed:
fix concurrency in MemoryLRUCache. change usage to TestUtils.tempDirectory()

dguy committed:
replace readwrite lock with synchronized in RocksDBStore as it could throw concurrent modification exceptions on get due to LRU eviction

guozhangwang commented:
One meta comment: we are currently exposing the following classes as public interfaces to users, under “state” sub-package:

QueryableStoreType
QueryableStoreTypes
ReadOnlyKeyValueStore
ReadOnlyWindowStore
StateStoreProvider

I’m wondering if we can move some of these classes into internals to minimize the number of user-facing interface classes? For example StateStoreProvider’s API currently is only used for state/internals classes.

dguy commented:
Good point @guozhangwang - i'll take a look at what can move. Thanks

dguy committed:
move StateStoreProvider to internals package. Remove test InMemoryKeyValueStore. make InMemoryStoreSupplier.MemoryStore thread-safe

guozhangwang commented:
@dguy Could you rebase?

dguy committed:
merge trunk

dguy commented:
@guozhangwang done

guozhangwang commented:
LGTM. Merged to trunk.
